File: EMUZ80_RP2040_PCB_MEZ68008_Firmware.py
#
# EMUZ80_RP2040_PCB_MEZ68008_Firmware
#
# @tendai22plus さんの EMUZ80_RP2040_PCB と
# @S_Okue さんの MEZ68008 を使って、
# Z80 DIP40 フットプリントに MC68008 を組み替えて動作させるための
# RP2040 ファームウェアです。
#
# 秋月電子通商 AE-RP2040 ボード上で動作。
# RP2040 が 64KB RAM + MC6850 互換 ACIA (0xE000) をエミュレート。
#
# Author:  DragonBallEZ
# License: MIT
#

from machine import Pin, PWM
import time
import select
import sys

def load_to_memory(memory, data, offset):
    """memoryの指定したoffset位置にdataをコピーする"""
    memory[offset : offset + len(data)] = data

adrs_pins = [Pin(i, Pin.IN, Pin.PULL_DOWN) for i in range(0, 16)]
data_pins = [Pin(i, Pin.IN, Pin.PULL_DOWN) for i in range(16, 24)]
as_pin = Pin(24, Pin.IN, Pin.PULL_UP) # GP24 in,  68k AS -  Z80 20pin IORQ
rw_pin = Pin(25, Pin.IN, Pin.PULL_UP) # GP25 in,  68k RW -  Z80 21pin RD
ds_pin = Pin(26, Pin.IN, Pin.PULL_UP) # GP26 in,  68k DS -  Z80 19pin MREQ
dtack_pin = Pin(27, Pin.OUT, value=0) # GP27 out, 68k DTACK - Z80 24pin WAIT
reset_pin = Pin(28, Pin.OPEN_DRAIN, value=0) # GP28 od, 68k RESET+HALT - Z80 26pin(RESET), 16pin(INT)
print("RESET-ON")


# UART
ACIAD = 0xE000
ACIAC = 0xE001

# Simulate microcontroller memory with bytearray
MEMORY_SIZE = 65536 # 64KB memory
memory = bytearray(MEMORY_SIZE)

test2 = bytearray([
    0x00, 0x01, 0x00, 0x00, 0x00, 0x00, 0x00, 0x08, 0x41, 0xFA, 0x00, 0x2C, 0x61, 0x00, 0x00, 0x1C, # 0x0000
    0x60, 0xFE, 0x48, 0x40, 0x10, 0x39, 0x00, 0x00, 0xE0, 0x01, 0xC0, 0x3C, 0x00, 0x02, 0x67, 0xF4, # 0x0010
    0x48, 0x40, 0x13, 0xC0, 0x00, 0x00, 0xE0, 0x00, 0x4E, 0x75, 0x10, 0x18, 0x67, 0x00, 0x00, 0x06, # 0x0020
    0x61, 0xE0, 0x60, 0xF6, 0x4E, 0x75, 0x0D, 0x0A, 0x20, 0x42, 0x79, 0x74, 0x65, 0x73, 0x20, 0x66, # 0x0030
    0x72, 0x65, 0x65, 0x0D, 0x0A, 0x0A, 0x45, 0x6E, 0x68, 0x61, 0x6E, 0x63, 0x65, 0x64, 0x20, 0x36, # 0x0040
    0x38, 0x6B, 0x20, 0x42, 0x41, 0x53, 0x49, 0x43, 0x20, 0x56, 0x65, 0x72, 0x73, 0x69, 0x6F, 0x6E, # 0x0050
    0x20, 0x33, 0x2E, 0x35, 0x34, 0x0D, 0x0A, 0x00                 # 0x0060
])

# アドレス 0x0000 から 0x0015 までのメモリイメージ（計22バイト）
# 0x0000-0x0003: 初期スタックポインタ (仮に 0x00001000 と設定)
# 0x0004-0x0007: 初期PC (プログラム開始アドレス 0x00000008)
# 0x0008-0x0015: 作成したループプログラム

test1 = bytearray([
    # ---- リセットベクトル領域 (0x0000 - 0x0007) ----
    0x00, 0x00, 0x10, 0x00,  # 0x0000: 初期SSP (0x00001000)
    0x00, 0x00, 0x00, 0x08,  # 0x0004: 初期PC  (0x00000008)
    
    # ---- プログラム領域 (0x0008 - 0x0015) ----
    0x41, 0xF8, 0x80, 0x00,  # 0x0008: LEA    $8000, A0
    0x10, 0x10,              # 0x000C: MOVE.B (A0), D0
    0x52, 0x00,              # 0x000E: ADDQ.B #1, D0
    0x10, 0x80,              # 0x0010: MOVE.B D0, (A0)
    0x4E, 0xF8, 0x00, 0x08   # 0x0012: JMP    $0008
])

@micropython.native
def busemu():
    # 標準入力(sys.stdin)の監視設定
    spoll = select.poll()
    spoll.register(sys.stdin, select.POLLIN)
    p_ds = 1
    while True:
        ds = ds_pin.value()
        if (p_ds == 1 and ds == 0):
#            values = [pin.value() for pin in adrs_pins]  # GPIO0-15 読取
#            adrs = sum(v << i for i, v in enumerate(values))  # 数値に変換 

            rw = rw_pin.value()
            if (rw == 1):
                if (adrs == ACIAC):
                    data = 0x02
    #                print("#")
                elif (adrs == ACIAD):
                    data = 0x41
    #                print("@")
                else:
                    data = memory[adrs]
                for i in range(8):
                    if (data & (1 << i)):
                        data_pins[i].value(1)
                    else:
                        data_pins[i].value(0)
                    data_pins[i].init(Pin.OUT)          
            else:
                values = [pin.value() for pin in data_pins]  # GPIO16-23 読取
                data = sum(v << i for i, v in enumerate(values))  # 数値に変換
                if (adrs == ACIAD):
#                    print("[" + chr(data) + "]", end="")
                    print(chr(data), end="")
                else:
                    memory[adrs] = data

#            print("AS:" + str(as_pin.value()), end=" ")
#            print("DS:" + str(ds_pin.value()), end=" ")
#            print("RW:" + str(rw), end=" ")
#            print(f"DATA: {data:02X} ADRS: {adrs:04X}")
        if (p_ds == 0 and ds_pin.value() == 1):
            if (rw_pin.value() == 1):
                for i in range(8):
                    data_pins[i].init(Pin.IN, Pin.PULL_DOWN)
        p_ds = ds
        # 待ち時間0ミリ秒で入力イベントをチェック
        events = spoll.poll(0)
        if events:
            # 入力があれば1文字読み込む
            char = sys.stdin.read(1)
            print(f"入力された文字: {char}")
            # 'q'が入力されたらループを抜ける
            if char == 'q':
                break
        time.sleep_ms(0)
    


load_to_memory(memory, test2, 0x0000)

lin = input("hit-any-key")
print("CLK-ON")
clk_pin = PWM(Pin(29), freq=170, duty_u16=32768) # GP29 out, 68k CLK - Z80 28pin RFSH
time.sleep(2)
reset_pin.value(1)
print("RESET-OFF")
print("Start..")
busemu()

clk_pin.deinit()

File: EMUZ80_RP2040_PCB_MEZ68008_Firmware.c
/*
 * EMUZ80_RP2040_PCB_MEZ68008_Firmware
 *
 * @tendai22plus さんの EMUZ80_RP2040_PCB と
 * @S_Okue さんの MEZ68008 を使って、
 * Z80 DIP40 フットプリントに MC68008 を組み替えて動作させるための
 * RP2040 ファームウェアです。
 *
 * 秋月電子通商 AE-RP2040 ボード上で動作。
 * RP2040 が 64KB RAM + MC6850 互換 ACIA (0xE000) をエミュレート。
 *
 * Author:  DragonBallEZ
 * License: MIT
 */

#include <stdio.h>
#include <string.h>
#include "pico/stdlib.h"
#include "hardware/gpio.h"
#include "hardware/pwm.h"
#include "hardware/clocks.h"
#include "hardware/structs/sio.h"
#include "pico/platform.h" 
#include "hardware/vreg.h"
#include "pico/multicore.h"
#include "tusb.h" // TinyUSB

#define ACIAD 0xE000
#define ACIAC 0xE001
#define MEMORY_SIZE 65536

// === Easy Overclock Configuration ===
// Change this one value to switch clock speed (under 300MHz recommended)
// Common good values: 200000, 250000, 266667, 280000
#define TARGET_SYS_CLK_KHZ 288000

// Pin assignments (EMUZ80_AE_RP2040 + MEZ68008 style)
#define ADDR_BASE 0
#define DATA_BASE 16
#define AS_PIN    24
#define RW_PIN    25
#define DS_PIN    26
#define DTACK_PIN 27
#define RESET_PIN 28
#define CLK_PIN   29

// Direct SIO masks for fast GPIO
#define DATA_MASK   (0xFFu << 16)   // GPIO16-23
#define DATA_SHIFT  16
#define DS_MASK     (1u << DS_PIN)
#define RW_MASK     (1u << RW_PIN)

#define DTACK_MASK (1u << DTACK_PIN)

// 68000 memory emulation area (64KB) aligned to 64KB boundary for performance
static uint8_t __attribute__((aligned(65536))) __not_in_flash("bus_memory")  memory[MEMORY_SIZE];

// Flag to indicate if the bus emulation is running (shared between cores)
static volatile bool    __attribute__((section(".scratch_x"))) emu_flg = false;  

#include "ehbasic.h"  // EhBASIC program image
// #include "ehbasic030.h"  // EhBASIC program image

static void load_to_memory(uint8_t *mem, const uint8_t *data, size_t len, uint16_t offset) {
    memcpy(mem + offset, data, len);
}

// Fast direct SIO register access (bypassing SDK for speed)
static inline void drive_data_fast(uint8_t data) {
    uint32_t bits = (uint32_t)data << DATA_SHIFT;
    sio_hw->gpio_out = (sio_hw->gpio_out & ~DATA_MASK) | bits;
    sio_hw->gpio_oe_set = DATA_MASK;
}

static inline void release_data_bus_fast(void) {
    sio_hw->gpio_oe_clr = DATA_MASK;
}

static inline uint32_t read_gpio_all(void) {
    return sio_hw->gpio_in;
}

// DTACK direct SIO control (for speed in busemu)
static inline void dtack_low(void)  { sio_hw->gpio_out = sio_hw->gpio_out & ~DTACK_MASK; }  // assert (active low)
static inline void dtack_high(void) { sio_hw->gpio_out = sio_hw->gpio_out |  DTACK_MASK; }  // deassert (high)

// 自動で最適な wrap / clkdiv を計算して PWM 周波数を設定する関数
static void pwm_set_frequency(uint slice_num, uint chan, float freq_hz) {
    uint32_t sys_hz = clock_get_hz(clk_sys);
    
    // 1周期あたりのシステムクロック数
    float period = (float)sys_hz / freq_hz;
    
    uint16_t wrap;
    float div;
    
    if (period > 65536.0f) {
        // 低周波数 → wrap最大 + 分周器を使う
        wrap = 0xFFFF;
        div = period / 65536.0f;
    } else {
        // 高周波数 → 分周器=1.0 で wrap を小さくする
        div = 1.0f;
        wrap = (uint16_t)(period - 1.0f);
        // 安全のため上限
        if (wrap > 0xFFFF) wrap = 0xFFFF;
    }
    
    pwm_config cfg = pwm_get_default_config();
    pwm_config_set_clkdiv(&cfg, div);
    pwm_config_set_wrap(&cfg, wrap);
    pwm_init(slice_num, &cfg, true);
    pwm_set_chan_level(slice_num, chan, wrap / 2);
    
    // 実際に出た周波数を表示（デバッグ用）
    float actual_freq = (float)sys_hz / ((wrap + 1ULL) * div);
    printf("PWM: target=%0.3fMHz  actual=%0.3fMHz  wrap=%u  div=%.4f\n",
           freq_hz / 1000000.0f, actual_freq / 1000000.0f, wrap, div);
}


// 通信専用のコア0,1で共有するUART通信のためのグローバル変数
static volatile uint8_t __attribute__((section(".scratch_x"))) uart_tx_data; // 送信データバッファ(Z80からPicoへ: 1バイト)
static volatile uint8_t __attribute__((section(".scratch_x"))) uart_rx_data; // 受信データバッファ(PicoからZ80へ: 1バイト)
static volatile bool __attribute__((section(".scratch_x"))) uart_tx_ready; // 送信可フラグ (true=Ready, false=Busy)
static volatile bool __attribute__((section(".scratch_x"))) uart_rx_ready; // 受信完了フラグ (true=Ready, false=Empty)

static volatile bool __attribute__((section(".scratch_x"))) exit_flag = false;


//
// コア1 のエントリポイント　バスエミュレーション
//
void __time_critical_func(busemu)(void) {
    uint8_t * const mem = memory;   // ベースアドレスを一度だけレジスタにロード
    uart_tx_data = 0;      // 送信データ初期化
    uart_rx_data = 0;      // 受信データ初期化
    uart_tx_ready = true;  // 送信準備完了
    uart_rx_ready = false; // 受信バッファは空
    exit_flag = false;     // 終了フラグ初期化

    while (!(read_gpio_all() & DS_MASK)) {  // DS low(active) → wait for high
        tight_loop_contents(); 
    }
    while (true) {
        while (read_gpio_all() & DS_MASK) { // DS high (inactive) → wait for falling edge
            tight_loop_contents(); 
        }
        uint32_t gpio = read_gpio_all();
        bool ds = (gpio & DS_MASK);   // ds = pin level (true when high/inactive)
        bool rw = (gpio & RW_MASK);
        uint16_t adrs = gpio & 0xFFFFu;
        uint8_t data;
        if (rw) {       // Read cycle: CPU is reading from us → drive data bus
            if (!((adrs & 0xFF00u) == 0xE000u)) {
                data = mem[adrs];
//                printf("A:%04X D:%02X\n", adrs, data);
            } else {
                if (adrs == ACIAC) {    // Status: bit0 (RDRF) = 1 if input char available
                    data = uart_rx_ready ? 0x01 : 0x00;  // RDRF
                    if (uart_tx_ready) {
                        data |= 0x02;   // TDRE set
                    }
                } else if (adrs == ACIAD) {
                    // Data: return char if available, else 0
                    data = uart_rx_ready ? uart_rx_data : 0;
                    uart_rx_ready = false;  // consume the char
                } else {
                    data = 0;
                }
            }
            drive_data_fast(data);
//            printf("RW=%d ADRS=%04X DATA=%02X\n", rw, adrs, data);
            
            dtack_low();  // DTACK low to acknowledge immediately
        } else {        // Write cycle
            data = (gpio >> DATA_SHIFT) & 0xFFu;
            if (adrs != ACIAD) {
                mem[adrs] = data;
            } else {
                if (uart_tx_ready) {       // 送信可能な場合のみ送信データを格納
                    uart_tx_data = data;
                    uart_tx_ready = false; // mark as busy
                } else {               // 送信中の場合は無視する（バッファオーバーフロー防止）
                }
            }
//            printf("RW=%d ADRS=%04X DATA=%02X\n", rw, adrs, data);
            dtack_low();  // DTACK low to acknowledge immediately
        }
        while (!(read_gpio_all() & DS_MASK)) {  // DS  low (active) → wait for rising edge
            tight_loop_contents();
        }
        if (rw) {            // RWを再サンプリングせず、立下り時にキャプチャしたフラグを使用
            release_data_bus_fast();
        }
        dtack_high();     // DTACK high to acknowledge immediately
        if (exit_flag) {
            break;  // exit on Ctrl-\ from UART
        }
    }
    emu_flg = false;  // signal to main that we are exiting
}

//
// --- UART Task (Core 0) ---
//
void uart_task(void) {
  printf("task UART start..\n\n");
  while (true) {
    // 送信処理: (Z80 -> USB) Z80がデータを書き込んで Busy になったら実行
    if (!uart_tx_ready) {
      if (tud_cdc_connected() && tud_cdc_write_available() > 0) {
        putchar(uart_tx_data);
        uart_tx_ready = true; // 送信完了（readyに戻す）
      }
    }
    // 受信処理(US->Z80) RX Readyがfalseの場合のみ入力をチェック
    if (!uart_rx_ready) {
      int c = getchar_timeout_us(0);
      if (c != PICO_ERROR_TIMEOUT) {
        if (c == 0x1C) { // Ctrl-\で終了
          exit_flag = true;
          break;
        }
        uart_rx_data = (uint8_t)c;
        uart_rx_ready = true;
      }
    }
    sleep_us(500); // 500us待機（CPU負荷を下げるため）
  }
}



int main() {
    // === Overclock setup (change TARGET_SYS_CLK_KHZ above to switch easily) ===
    uint32_t target_khz = TARGET_SYS_CLK_KHZ;

    // Select safe voltage based on target frequency
    if (target_khz <= 200000) {
        vreg_set_voltage(VREG_VOLTAGE_1_15);
    } else if (target_khz <= 250000) {
        vreg_set_voltage(VREG_VOLTAGE_1_25);
    } else {
        vreg_set_voltage(VREG_VOLTAGE_1_30);
    }

    sleep_ms(5);  // wait for voltage to stabilize

    bool clock_ok = set_sys_clock_khz(target_khz, true);

    stdio_init_all();
    sleep_ms(2000);

    printf("EMUZ80_RP2040 MEZ68008 - 0.01\n");
    if (clock_ok) {
        printf("sysclk %0.3fMHz\n", target_khz / 1000.0f);
    } else {
        printf("Failed to set %0.3fMHz clock\n", target_khz / 1000.0f);
    }

    // Address bus: GP0-15 inputs with pull-down
    for (int i = 0; i < 16; i++) {
        gpio_init(ADDR_BASE + i);
        gpio_pull_down(ADDR_BASE + i);
        gpio_set_dir(ADDR_BASE + i, GPIO_IN);
    }

    // Data bus: GP16-23 inputs with pull-down (bidirectional)
    for (int i = 0; i < 8; i++) {
        gpio_init(DATA_BASE + i);
        gpio_pull_down(DATA_BASE + i);
        gpio_set_dir(DATA_BASE + i, GPIO_IN);
    }
    // Make sure data bus is input from SIO perspective
    sio_hw->gpio_oe_clr = DATA_MASK;

    gpio_init(AS_PIN);
    gpio_pull_up(AS_PIN);
    gpio_set_dir(AS_PIN, GPIO_IN);

    gpio_init(RW_PIN);
    gpio_pull_up(RW_PIN);
    gpio_set_dir(RW_PIN, GPIO_IN);

    gpio_init(DS_PIN);
    gpio_pull_up(DS_PIN);
    gpio_set_dir(DS_PIN, GPIO_IN);

    gpio_init(DTACK_PIN);
    gpio_put(DTACK_PIN, 1);  // DTACK初期値はHIGH
    gpio_set_dir(DTACK_PIN, GPIO_OUT);

    // RESET (open-drain style): start asserted
    gpio_init(RESET_PIN);
    gpio_put(RESET_PIN, 0);
    gpio_set_dir(RESET_PIN, GPIO_OUT);
    printf("RESET-ON\n");

  
    printf("hit-any-key\n");
    while (getchar_timeout_us(0) == PICO_ERROR_TIMEOUT) {
        tight_loop_contents();  // yield a bit
    }
    // Start CLK via PWM on GP29 (自動計算で周波数設定)
    printf("CLK-ON\n");
    gpio_set_function(CLK_PIN, GPIO_FUNC_PWM);
    uint slice_num = pwm_gpio_to_slice_num(CLK_PIN);
    uint chan = pwm_gpio_to_channel(CLK_PIN);

   float desired_freq = 10000000.0f;    // 10MHz
//    float desired_freq = 8000000.0f;    // 8MHz

   pwm_set_frequency(slice_num, chan, desired_freq);
    sleep_ms(200);

    printf("ROMエミュレータ起動 - core1\n");
    multicore_launch_core1(busemu);
    emu_flg = true;

    // Load program into simulated memory EhBASIC
    load_to_memory(memory, EhBASIC, ehbasic_size, 0x0000);
    printf("EhBASIC loaded to memory (size=%d bytes)\n", ehbasic_size);
    // Wait for any keypress to proceed

    printf("RESET-OFF\n");
    printf("Start..\n");
    // Release reset
    gpio_set_dir(RESET_PIN, GPIO_IN);   // RESET-OFF (open-drain style)
    uart_task();  // Start UART task on core 0

    pwm_set_enabled(slice_num, false);

    printf("\nStopped.\n");

    while (true) {
        tight_loop_contents();
    }
}
